### DISCUSSION TITLE
Awkward arrays in C++ / ROOT

### DISCUSSION THREAD

**Discussion by dinardo** [scikit-hep/awkward#2470] [General]
Dear Experts,
I came across this interesting paper: https://arxiv.org/pdf/2303.02205.pdf
It's not clear to me whether I can create an awkward array in C++ and "convert it" into a `TTree`, avoiding the annoying ROOT dictionary generation procedure.

My final goal would be to have a `TTree` branch which is a: `std::vector< std::vector<uint32_t> >`

Many thanks.
- Mauro.

**Comment by agoose77** [comment]
Hi @dinardo!

Awkward Array is not intended to be used exclusively from C++. The paper that you referenced describes a mechanism for creating Awkward Arrays in C++; this header-only library is intended for use in C++ libraries that expose a Python API, but it is of course possible to write these arrays to e.g. disk, and load them in a separate process.

We implement a `to_rdataframe` API function that supports interaction with Awkward Arrays in an RDataFrame context. If your array is only singly-jagged, i.e. the array `x` shown in [this documentation](https://awkward-array.org/doc/main/reference/generated/ak.to_rdataframe.html), then you can `Snapshot` the dataframe to disk to build a TTree. However, if the type is more complex than that, this approach cannot be used.

Instead, there is also `uproot`, which provides a mechanism for building new `TTree`s from many types, including Awkward Arrays. Of course, this requires that you use Python to use these libraries (though you can even do this _from_ C++, e.g. with `pybind11` for convenience).

**Reply by dinardo** [reply]
Many thanks @agoose77 ,
it's all clear.
Unfortunately, my array is not "singly jagged", because it's "an array of an array".
Ragards,
- Mauro.

**Reply by jpivarski** [reply]
If by "array of an array," you mean an array of (variable-length) lists of numbers, like this

```python
>>> content = [0.0, 1.1, 2.2, 3.3, 4.4, 5.5, 6.6, 7.7, 8.8, 9.9]
>>> lengths1 = [3, 0, 2, 1, 4]
>>> ak.unflatten(content, lengths1).show()
[[0, 1.1, 2.2],
 [],
 [3.3, 4.4],
 [5.5],
 [6.6, 7.7, 8.8, 9.9]]
```

then that is what we mean by a singly-jagged array. But also, [ak.from_rdataframe](https://awkward-array.org/doc/main/reference/generated/ak.from_rdataframe.html) and Uproot can convert from an arbitrarily deep nesting of lists of lists of lists (i.e. doubly-jagged, triply-jagged):

```python
>>> lengths2 = [2, 0, 3]
>>> ak.unflatten(ak.unflatten(content, lengths1), lengths2).show()
[[[0, 1.1, 2.2], []],
 [],
 [[3.3, 4.4], [5.5], [6.6, 7.7, 8.8, 9.9]]]
```

Oh! @agoose77 is right: he's saying that if you pass an Awkward Array into an RDataFrame with [ak.to_rdataframe](https://awkward-array.org/doc/main/reference/generated/ak.to_rdataframe.html), it won't be able to `Snapshot` it unless it's only one level deep (singly-jagged). [Uproot has the same limitation.](https://uproot.readthedocs.io/en/latest/basic.html#writing-ttrees-to-a-file)

**Comment by dinardo** [comment]
Hi @jpivarski ,
we might have a word-meaning overload :-)

I need to make a `TTree` branch of: `std::vector< std::vector<uint32_t> >` in C++ (none of the two vectors has a predefined length)
I'm not using `python`

Normally `ROOT` requires that I make a `dictionary` which I would like to avoid.
I was wondering whether I could somehow define my structure in `awkward` and convert it into a `TTree`

I must say that I'm not super familiar with all these new packages, but I keep reading around about "conversion in RDataFrame"
As far as I understood `RDataFrame` is just an interface to a `TTree`, the underlying data structure is actually a `TTree`, in fact, I can't define `RDataFrame` "branches" and fill them one by one as if they were HEP events. But I might got it all wrong.

Many thanks.
- Mauro.

**Comment by NJManganelli** [comment]
Hi @dinardo,

That would be "double-jagged" data in this scheme, since the TTree Entry index is always treated as knowable/findable/un-fixed. So a variable number of objects per TTree Entry is single-jagged, and variable number of variable number of things is doubly-jagged.

RDataFrame is, in a single blurb, a "Declarative interface to TTree (and other) data." It has Snapshot functionality to write out TTrees (and through the declarative Define calls you can create new quantities, and use Filter calls to make selections on TTree entries), but it doesn't have generalized (automatic?) support for more than 1 level of jaggedness right now. That is, it works perfectly fine for CMS NanoAOD and similar ntuples.

Since the interface is declarative, you don't "manually" fill things event-by-event, but the end result is similar (within its constraints). 

[https://root.cern/doc/master/classROOT_1_1RDataFrame.html](https://root.cern/doc/master/classROOT_1_1RDataFrame.html) shows some examples.

Best,
Nick

**Reply by dinardo** [reply]
Hi,
ok many thanks for your clarification.
It's more clear now.
Cheers,
- Mauro.
